### PR TITLE
Own http headers

### LINKED ISSUE #45: Own HTTP Headers
The proposed structure is:

```swift
struct HTTP {
    struct Header {
        struct Key {
            /// here are the keys for headers
        }
    }
}
```

and then have headers with type:
```swift
let headers: [HTTP.Header.Key: String]
```

### PR DESCRIPTION
**Pull request by syllerim**
Closes #45

### CODE REVIEW

**Review by alickbass — CHANGES_REQUESTED** [review]
Please resolve git conflicts by rebasing your branch on top of master

File: ApiConnectorTests/TestHTTP.swift [hunk 1]
```diff
@@ -0,0 +1,45 @@
...
+
+class TestHTTP: XCTestCase {
+    
+    func testHTTPHashable() {
+        XCTAssertEqual(HTTP.Header.Key.accept, "Accept")
+        XCTAssertNotEqual(HTTP.Header.Key.accept, "Accept-Language")
+        XCTAssertEqual(HTTP.Header.Key.accept.hashValue, HTTP.Header.Key("Accept").hashValue)
+    }
+    
+    func testHTTHeader() {
```
Comment (alickbass) [inline ApiConnectorTests/TestHTTP.swift]: To be honest, I think that this test is a bit useless, as it tests `Dictionary`, which we know is working. So I think that you can remove completely this method

File: ApiConnectorTests/TestHTTP.swift [hunk 2]
```diff
@@ -0,0 +1,45 @@
...
+        XCTAssertEqual(headers[.contentType], result[HTTP.Header.Key.contentType.rawValue])
+        
+    }
+    
+    func testHTTPToString() {
+        let headers: [HTTP.Header.Key: String] = [.accept: "application/json" ,
+                                                  .acceptLanguage: "en-US",
+                                                  .contentType: "charset=utf-8"]
+        var result: [String: String]
+        result = HTTP.Header.toStringKeys(headers: headers)
```
Comment (alickbass) [inline ApiConnectorTests/TestHTTP.swift]: You can just write
```swift
let result = HTTP.Header.toStringKeys(headers: headers)
```

File: ApiConnectorTests/TestHTTP.swift [hunk 3]
```diff
@@ -0,0 +1,45 @@
...
+        XCTAssertEqual(headers.count, result.count)
+        XCTAssertEqual(headers[.accept], result[HTTP.Header.Key.accept.rawValue])
+        XCTAssertEqual(headers[.acceptLanguage], result[HTTP.Header.Key.acceptLanguage.rawValue])
+        XCTAssertEqual(headers[.contentType], result[HTTP.Header.Key.contentType.rawValue])
```
Comment (alickbass) [inline ApiConnectorTests/TestHTTP.swift]: I would use the result from previous method:
```swift 
let result = [HTTP.Header.Key.accept.rawValue: "application/json" ,
                     HTTP.Header.Key.acceptLanguage.rawValue: "en-US",
                     HTTP.Header.Key.contentType.rawValue: "charset=utf-8"]
```

and then just compared 2 dictionaries:

```swift
XCTAssertEqual(result, HTTP.Header.toStringKeys(headers: headers))
```

File: ApiConnector/HTTP.swift [hunk 4]
```diff
@@ -0,0 +1,78 @@
...
+    public static let contentType  = HTTP.Header.Key("Content-Type")
+}
+
+extension HTTP.Header.Key : ExpressibleByStringLiteral {
+    public init(stringLiteral value: String) {
+        self.init(value)
+    }
+    
+    public init(extendedGraphemeClusterLiteral value: String){
+        self.init(stringLiteral: value)
```
Comment (alickbass) [inline ApiConnector/HTTP.swift:62]: This method is not covered with tests. For the example of how to cover such cases, take a look [here](https://github.com/Jitsusama/UInt128/blob/master/tests/UInt128Tests.swift) at line 147

File: ApiConnector/HTTP.swift [hunk 5]
```diff
@@ -0,0 +1,78 @@
...
+    }
+    
+    public init(unicodeScalarLiteral value: String) {
+        self.init(stringLiteral: value)
```
Comment (alickbass) [inline ApiConnector/HTTP.swift:66]: This method is not covered with tests. For the example of how to cover such cases, take a look [here](https://github.com/Jitsusama/UInt128/blob/master/tests/UInt128Tests.swift) at line 141

**Review by alickbass — CHANGES_REQUESTED** [review]

File: ApiConnectorTests/TestHTTP.swift (hunk 2 above)
Comment (syllerim) [reply ApiConnectorTests/TestHTTP.swift]: yeah, this is nonsense

File: ApiConnectorTests/TestHTTP.swift [hunk 6]
```diff
@@ -0,0 +1,45 @@
...
+                      HTTP.Header.Key.contentType.rawValue: "charset=utf-8"]
+        XCTAssertEqual(headers.count, result.count)
+        XCTAssertEqual(headers[.accept], result[HTTP.Header.Key.accept.rawValue])
+        XCTAssertEqual(headers[.acceptLanguage], result[HTTP.Header.Key.acceptLanguage.rawValue])
+        XCTAssertEqual(headers[.contentType], result[HTTP.Header.Key.contentType.rawValue])
+        
+    }
+    
+    func testHTTPToString() {
+        let headers: [HTTP.Header.Key: String] = [.accept: "application/json" ,
```
Comment (alickbass) [inline ApiConnectorTests/TestHTTP.swift]: BTW, here you can do like this:
```swift
let headers: HTTP.Headers = ....
```

File: ApiConnectorTests/ApiConnectionTests.swift [hunk 7]
```diff
@@ -29,13 +29,17 @@ fileprivate class TestValidationConnection<Provider: ResponseProvider>: ApiConne
 class ApiConnectionTests: XCTestCase {
     
     func testRequestCreation() {
+        let header: [HTTP.Header.Key: String] = [.accept: "application/json"]
```
Comment (alickbass) [inline ApiConnectorTests/ApiConnectionTests.swift]: Here use can also change to:

```swift
let headers: HTTP.Headers = ...
```

File: ApiConnectorTests/ApiConnectionTests.swift [hunk 8]
```diff
@@ -29,13 +29,17 @@ fileprivate class TestValidationConnection<Provider: ResponseProvider>: ApiConne
...
-            .request(method:.post, with: TestData.testBodyData, at: .me, headers: nil)
-        var expectedRequest = try! URLRequest(url: Router<Api>(.test, at: .me).url, method: .post, headers: nil)
-        expectedRequest.httpBody = TestData.testBodyData
+            .request(method:.post, with: TestData.testBodyData, at: .me, headers: header)
         
+        var expectedRequest = try! URLRequest(url: Router<Api>(.test, at: .me).url, method: .post, headers: headerToString)
+        expectedRequest.httpBody = TestData.testBodyData
         XCTAssertEqual(request, expectedRequest)
         XCTAssertEqual(request.httpMethod, HTTP.Method.post.rawValue)
+        XCTAssertEqual(request.allHTTPHeaderFields?[HTTP.Header.Key.accept.rawValue], headerToString[HTTP.Header.Key.accept.rawValue])
```
Comment (alickbass) [inline ApiConnectorTests/ApiConnectionTests.swift]: Here you can just do:
```swift
XCTAssertEqual(request.allHTTPHeaderFields!, headersToString)
```

**Comment by alickbass** [comment]
Can you please rename the file from `TestHTTP` to `HTTPTests` to be consistent with other tests

File: ApiConnector/HTTP.swift (hunk 4 above)
Comment (syllerim) [reply ApiConnector/HTTP.swift:62]: Implemented

File: ApiConnector/HTTP.swift (hunk 5 above)
Comment (syllerim) [reply ApiConnector/HTTP.swift:66]: Implemented

File: ApiConnectorTests/HTTPTest.swift [hunk 9]
```diff
@@ -0,0 +1,35 @@
...
+                                     .contentType: "charset=utf-8"]
+        let result = [HTTP.Header.Key.accept.rawValue: "application/json" ,
+                      HTTP.Header.Key.acceptLanguage.rawValue: "en-US",
+                      HTTP.Header.Key.contentType.rawValue: "charset=utf-8"]
+         XCTAssertEqual(result, HTTP.Header.toStringKeys(headers: headers))
+    }
+    
+    func testHTTPHeadersStringLiteral(){
+        XCTAssertEqual(HTTP.Header.Key(unicodeScalarLiteral: "김").rawValue, "김")
+        XCTAssertEqual(HTTP.Header.Key(extendedGraphemeClusterLiteral: "22").rawValue, "\u{00032}\u{00032}")
```
Comment (alickbass) [inline ApiConnectorTests/HTTPTest.swift]: Should be the other way around here:
```swift
HTTP.Header.Key(extendedGraphemeClusterLiteral: "\u{00032}\u{00032}").rawValue, "22"
```

File: ApiConnectorTests/HTTPTest.swift [hunk 10]
```diff
@@ -0,0 +1,35 @@
+//
+//  TestHTTP.swift
```
Comment (alickbass) [inline ApiConnectorTests/HTTPTest.swift]: Can you please rename the file t `HTTPTests` instead of `HTTPTest`

**Review by alickbass — CHANGES_REQUESTED** [review]

File: ApiConnectorTests/HTTPTest.swift (hunk 10 above)
Comment (syllerim) [reply ApiConnectorTests/HTTPTest.swift]: sure

File: ApiConnectorTests/HTTPTest.swift (hunk 9 above)
Comment (syllerim) [reply ApiConnectorTests/HTTPTest.swift]: sure

**Review by alickbass — APPROVED** [review]

**Pull request merged**